### PR TITLE
Add support for $releasever_major, $releasever_minor, and limited shell expansion

### PR DESCRIPTION
**Pull request by evan-goode**
Resolves https://github.com/rpm-software-management/dnf5/issues/710.

Add `$releasever_major` and `$releasever_minor`: whenever the `releasever` var is set, automatically derive and set the `releasever_major` and `releasever_minor` vars by splitting `releasever` on the first dot. In the future (or now, possibly), we may also want to add `releasever_patch`.

Compatible with this implementation in libzypp:
https://github.com/openSUSE/libzypp/blob/d03a02d986f43266af81c73cced1b310e8615971/zypp/repo/RepoVariables.cc#L478-L495

Also adds support for `${variable:-word}` and `${variable:+word}` expansions of vars, similar to the functionality in POSIX shell.

`${variable:-word}` means if `variable` is unset or empty, the expansion of `word` is substituted. Otherwise, the value of `variable` is substituted.

`${variable:+word}` means if `variable` is unset or empty, nothing is substituted. Otherwise, the expansion of `word` is substituted.

Both support nested variable expressions; you could do `${var0:-${var1:+${var2:+$var3}}}`

Zypper also supports these expansions, see here: https://doc.opensuse.org/projects/libzypp/HEAD/structzypp_1_1repo_1_1RepoVarExpand.html

### CODE REVIEW

File: include/libdnf5/conf/vars.hpp [hunk 1]
```diff
@@ -86,7 +95,13 @@ struct Vars {
     /// @param name Name of the variable
     const Variable & get(const std::string & name) const { return variables.at(name); }
 
-    static std::unique_ptr<std::string> detect_release(const BaseWeakPtr & base, const std::string & install_root_path);
+    static std::optional<std::string> detect_release(const BaseWeakPtr & base, const std::string & install_root_path);
```
Comment (j-mracek) [inline include/libdnf5/conf/vars.hpp:98]: I am sorry, but we cannot use std::optional because it is not supported by SWIG.

File: include/libdnf5/conf/vars.hpp [hunk 2]
```diff
@@ -70,6 +71,14 @@ struct Vars {
     /// @param prio Source/Priority of the value
     void set(const std::string & name, const std::string & value, Priority prio = Priority::RUNTIME);
 
+    /// @brief Set a variable to a value, only obtaining the value if needed using `get_value`
+    ///
+    /// @param name Name of the variable
+    /// @param get_value Function that returns the (optional) value for the variable
+    /// @param prio Source/Priority of the value
+    void set_lazy(
```
Comment (j-mracek) [inline include/libdnf5/conf/vars.hpp:79]: I am not sure whether it is a good idea to detect vars not at the same time point. Right not it happen at Base::setup() which is a mandatory step. Is there a particular reason to do it this way? Originally we were thinking about locking vars like with configuration, but the feature has low priority therefore it is not yet implemented.

File: include/libdnf5/conf/vars.hpp (hunk 1 above)
Comment (evan-goode) [reply include/libdnf5/conf/vars.hpp:98]: Ah, forgot about that, that's a shame. I'll use `unique_ptr` instead.

File: include/libdnf5/conf/vars.hpp (hunk 2 above)
Comment (evan-goode) [reply include/libdnf5/conf/vars.hpp:79]: `set_lazy` doesn't actually defer setting the var, it's the same as `set` but only evaluates the value to be set if it's necessary to do so, i.e. when the incoming priority is greater than the existing priority.

It's a somewhat common pattern, often you will see a `log.Println(() => string)` function that only evaluates the lambda if necessary.

I added it to deduplicate some of the code in `detect_vars`.

File: libdnf5/conf/vars.cpp [hunk 3]
```diff
@@ -223,17 +223,52 @@ std::string Vars::substitute(const std::string & text) const {
...
+    if (it != variables.end() && prio < it->second.priority) {
+        return;
+    }
+
+    // Whenever releasever is set, split it into major and minor parts
+    if (name == "releasever") {
+        const auto [releasever_major, releasever_minor] = split_releasever(value);
+        set("releasever_major", releasever_major, prio);
+        set("releasever_minor", releasever_minor, prio);
+    }
```
Comment (j-mracek) [inline libdnf5/conf/vars.cpp:345]: I am not sure whether automatic detection during the setting is the right approach. Anyway it is not wrong.

I have some concerns to priority and setting empty values.
The current code using the same priority for derived variables. Why cannot we use a lover priority? Example user define from commandline `releasever_major` and then `releasever` but it overrides `releasever_major` because autodetected value has the same priority.

With empty value I guess that we should not substitute `$releasever_minor` when `releasever` = f38. DNF does that in other places (when no releasever (in installroot), it does not substitute it) and I somehow feel it  would be good to keep the same pattern.  This behavior is different from shell, but it helps with debugging.

File: include/libdnf5/conf/vars.hpp (hunk 2 above)
Comment (j-mracek) [reply include/libdnf5/conf/vars.hpp:79]: Thank you very much for explanation. Do you think we need it as API?

File: libdnf5/conf/vars.cpp (hunk 3 above)
Comment (m-blaha) [reply libdnf5/conf/vars.cpp:345]: Hmm, playing with priorities and setting `releasever_major` with artificially lowered priority doesn't seem right to me. I believe the root issue here lies in the design itself - `releasever_major` and `releasever_minor` are dependent on the value of the `releasever` variable and should be treated as kind of  "readonly" (I'd even consider disallowing independent setting, allowing changes only through `releasever`).

Comment (j-mracek) [reply libdnf5/conf/vars.cpp:368]: That would be possible, but it would require to implement locking mechanism for Vars, like we have for configuration and it will create an exception in the logic.

Comment (evan-goode) [reply libdnf5/conf/vars.cpp]: > Hmm, playing with priorities and setting `releasever_major` with artificially lowered priority doesn't seem right to me. I believe the root issue here lies in the design itself - `releasever_major` and `releasever_minor` are dependent on the value of the `releasever` variable and should be treated as kind of "readonly" (I'd even consider disallowing independent setting, allowing changes only through `releasever`).

I agree, I think this is the better approach, I made `releasever_major` and `releasever_minor` read-only. Let me know what you think.

File: libdnf5/conf/vars.cpp [hunk 4]
```diff
@@ -35,7 +35,6 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include <cstring>
 #include <filesystem>
 #include <memory>
-#include <optional>
```
Comment (Conan-Kudo) [inline libdnf5/conf/vars.cpp:38]: Why is `optional` getting removed here? It seems independent from the rest of the changes? Can we split this to a separate commit at least?

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

File: libdnf5/conf/vars.cpp (hunk 4 above)
Comment (evan-goode) [reply libdnf5/conf/vars.cpp]: Whoops, thanks. I've put it back in.

**Comment by Conan-Kudo** [comment]
@evan-goode Can you rebase this to get rid of the conflicts?

**Comment by evan-goode** [comment]
> @evan-goode Can you rebase this to get rid of the conflicts?

Yep, rebased.

File: include/libdnf5/conf/vars.hpp (hunk 2 above)
Comment (Conan-Kudo) [reply include/libdnf5/conf/vars.hpp:78]: You could just move it to the `private` section of the class if it's only used in the Vars class.

Comment (evan-goode) [reply include/libdnf5/conf/vars.hpp]: OK, done.

File: libdnf5/conf/vars.cpp (hunk 3 above)
Comment (Conan-Kudo) [reply libdnf5/conf/vars.cpp]: I think this sufficiently addressed this feedback.

Comment (Conan-Kudo) [reply libdnf5/conf/vars.cpp]: I think this sufficiently addresses the feedback. And this is the only feedback blocking me from merging this.

@j-mracek are we good with this now?

File: include/libdnf5/conf/vars.hpp [hunk 5]
```diff
@@ -88,6 +88,12 @@ struct Vars {
 
     static std::unique_ptr<std::string> detect_release(const BaseWeakPtr & base, const std::string & install_root_path);
 
+    /// @brief Split releasever on the first "." into its "major" and "minor" components
+    ///
+    /// @param releasever A releasever string, possibly containing a "."
+    /// @return releasever_major, releasever_minor
+    static std::tuple<std::string, std::string> split_releasever(const std::string & releasever);
```
Comment (j-mracek) [inline include/libdnf5/conf/vars.hpp:101]: Be honest I would not prefer to add the method to API and make it private. If we will add support of releasever_patch value, the method will be obsoleted.

File: libdnf5/conf/vars.cpp [hunk 6]
```diff
@@ -180,67 +182,243 @@ std::unique_ptr<std::string> Vars::detect_release(const BaseWeakPtr & base, cons
...
+    return std::make_tuple(releasever_major, releasever_minor);
+}
+
+bool Vars::is_read_only(const std::string & name) {
+    return READ_ONLY_VARIABLES.contains(name);
 }
 
 void Vars::set(const std::string & name, const std::string & value, Priority prio) {
+    if (is_read_only(name)) {
+        throw RuntimeError(M_("Variable \"{}\" is read-only"), name);
```
Comment (j-mracek) [inline libdnf5/conf/vars.cpp:377]: Please document that it can throw an error. I have to test, but I think that it will create a problem with setvars option or other channels that can modify variables.

Comment (j-mracek) [reply libdnf5/conf/vars.cpp:377]: I've tested it and it works like expected. I am still not sure whether RuntimeError is the best exception.

Comment (j-mracek) [reply libdnf5/conf/vars.cpp:377]: I know that it is used in the rest of the module but I feel that libdnf5 exception that inherits from runtime would be more appropriate. What do you think?

File: include/libdnf5/conf/vars.hpp (hunk 5 above)
Comment (evan-goode) [reply include/libdnf5/conf/vars.hpp]: Done, makes sense. An API user could always read the value of `releasever_major`, `releasever_minor` after setting `releasever`.

File: libdnf5/conf/vars.cpp (hunk 6 above)
Comment (evan-goode) [reply libdnf5/conf/vars.cpp]: I agree, I added a `ReadOnlyVariableError` class, and I put a `/// @throw ReadOnlyVariableError` in the docstring. Let me know if I should document it anywhere else.

I added a test here https://github.com/rpm-software-management/ci-dnf-stack/pull/1359 to ensure read-only variables cannot be set.

Comment (j-mracek) [reply libdnf5/conf/vars.cpp]: Additionally I asked @jrohel about opinion for locking mechanism, because he is an author of config locking mechanism.

Comment (jrohel) [reply libdnf5/conf/vars.cpp]: This is not about locking. This code implements a hard-coded list of read-only variable names. The value of these variables is not locked, but is calculated based on the value of another variable. In other words, the value of these variables is not locked, they can be changed at any time, indirectly, by changing the value of another variable.

File: libdnf5/conf/vars.cpp [hunk 7]
```diff
@@ -179,54 +180,170 @@ std::unique_ptr<std::string> Vars::detect_release(const BaseWeakPtr & base, cons
...
-
 std::string Vars::substitute(const std::string & text) const {
-    std::string res = text;
+    return substitute_expression(text, 0).first;
+}
 
-    if (variables.empty()) {
-        return res;
+const unsigned int MAXIMUM_EXPRESSION_DEPTH = 32;
+std::pair<std::string, size_t> Vars::substitute_expression(const std::string_view & text, unsigned int depth) const {
```
Comment (jrohel) [inline libdnf5/conf/vars.cpp:192]: Please always pass `std::string_view` by value. 
`const std::string_view & text` -> `std::string_view text `

Comment (evan-goode) [reply libdnf5/conf/vars.cpp]: Done

File: include/libdnf5/conf/vars.hpp [hunk 8]
```diff
@@ -68,8 +76,15 @@ struct Vars {
...
     /// @param value Value to be stored in variable
     /// @param prio Source/Priority of the value
+    /// @throw ReadOnlyVariableError if the variable is read-only
     void set(const std::string & name, const std::string & value, Priority prio = Priority::RUNTIME);
 
+    /// @brief Checks whether a variable is read-only
+    ///
+    /// @param name Name of the variable
+    /// @return true if the variable is read-only, false if it is writable
+    bool is_read_only(const std::string & name);
```
[the remark this answers is not in the document]
Comment (evan-goode) [reply include/libdnf5/conf/vars.hpp]: Done
